### ISSUE TITLE
[stow task] color histogram matching 

### ISSUE THREAD

**Issue by knorth55** [start-jsk/jsk_apc#1050]
@kimhc6028 
HSI color histogram matching 

https://en.wikipedia.org/wiki/Histogram_matching
https://en.wikipedia.org/wiki/Histogram_equalization
http://stackoverflow.com/questions/19907822/histogram-equalization-based-on-hsi-model

**Comment by knorth55** [comment]
put more info for improvement

**Comment by kimhc6028** [comment]
![hinomaru4png](https://cloud.githubusercontent.com/assets/15006217/12971683/299c5e2e-d0e1-11e5-95e8-93eeb02c6e12.png)
![korea_flag0png](https://cloud.githubusercontent.com/assets/15006217/12971684/299e9fe0-d0e1-11e5-8e65-04f6e707c6ff.png)
Extracted Hue and Gray color histogram from Hinomaru and Korean Flag. However, it seems like ordinary classifier(probably SVM?) could not classify well from this data.

**Comment by k-okada** [comment]
nice work, do you know how to learn apc objects? if not please ask @wkentaro

--
◉ Kei Okada

**Comment by kimhc6028** [comment]
Only if I can prepare Berkeley data set and extract masked image from it, I expect I can. I will ask wkentaro.
By the way, test result for distinguishing korea_flag and hinomaru was like this
'''
header: 
  seq: 23
  stamp: 
    secs: 1455207374
    nsecs: 730146884
  frame_id: camera
labels: [0]
label_names: ['Hinomaru']
label_proba: [0.5137508264999706]
probabilities: [0.5137508264999706, 0.4862491735000293]
classifier: <jsk_2015_05_baxter_apc.ColorHistogramFeatures>
target_names: ['Korea_Flag', 'Hinomaru']
'''
The probability difference between two classes were too small. I think SVM classifier is not good enough for this classification task. Thus, I am thinking of adopting simple neural network. How do you think of it, sir?

**Comment by wkentaro** [comment]
> Only if I can prepare Berkeley data set and extract masked image from it, I expect I can. I will ask wkentaro.

The way to prepare dataset is written here: 
https://github.com/start-jsk/jsk_apc/tree/master/jsk_apc2015_common#train-classifier-with-dataset

> The probability difference between two classes were too small. I think SVM classifier is not good enough for this classification task.

I think the considerable reasons are:

- too small dataset
- color histogram feature is not enough (it ignores features coords data)

that's why I combined (boosted) CH(Color Histogram) and SIFT-BOF(Bag of Features with SIFT) based classifiers at APC2015.

>  I am thinking of adopting simple neural network. 

Yeah, I agree with that if you have time.
I'm also trying the object detection task with convolutional-neural-network.

**Comment by wkentaro** [comment]
I shared the dataset as described at #1060

**Comment by kimhc6028** [comment]
I have tested accuracy of color histogram algorithm.
Training data was berkeley dataset, and test dataset was 50 pictures I took personally. Because lack of test dataset, classification for only 7 classes was done.
I tested SVM and neural network for classifier.
Also, 3 kinds of color histograms were tested.
1) RGB color space
2) Only Hue data on HSV color space
3) Hue and grayscale data on HSV color scale
The result was
'''
SVM, RGB -> 0.58
SVM, Hue & Gray -> 0.72
SVM, Hue -> 0.78
Neural Network, RGB -> 0.46
Neural Network, Hue & Gray -> 0.58
Neural Network, Hue -> 0.76
'''
This indicates using hue histogram increases its performance. I guess it is because hue histogram is not affected by brightness of images.

**Comment by k-okada** [comment]
which algorithm is used in current RGB based clustering? SVM? NN? or others?

--
◉ Kei Okada

**Comment by kimhc6028** [comment]
Current RGB based clustering uses SVM.

**Comment by k-okada** [comment]
Ok, so we'd like to see if this works with stow settings which we struggled last week

**Comment by kimhc6028** [comment]
We have tested new histogram(hue only) with Baxter's vision. However, it totally failed to classify objects at all. i.e , this classify oreo as rolodex_cup (which is totally black one). 

In addition, I tested classifier's accuracy with test dataset I made 3 days ago. Unlike experiment conducted 3 days ago, this time SVM classifier is trained with all 25 objects. This shows dramatic decrease of accuracy. With hue only histogram, it shows 50% accuracy. With previous RGB histogram, it show 22% accuracy.
It might show color histogram is not enough at all.

Also, I think there is problem on masked dataset. Masking uses bounding_rect_of_mask program on jsk_recognition_utils, which generates rectangular mask. This rectangular mask includes background, which is critical for color histogram.
![n1_258](https://cloud.githubusercontent.com/assets/15006217/13075551/800166d0-d4ef-11e5-8168-fab1f225f152.jpg)
Thus, mask generation algorithm should be improved, generate masked image with transparent masking like this ;
![cat](https://cloud.githubusercontent.com/assets/15006217/13076299/c867db02-d4f4-11e5-9731-11532e3f7d6f.jpg)

**Comment by kimhc6028** [comment]
@knorth55 
How about visualize color histogram of masked image of objects in bins, and compare it with color histogram of training images?
I committed visualization program (node_scripts/histogram_plot.py) on my color_object_features branch.

**Comment by k-okada** [comment]
I think we already histogram visualization in somewhere in jsk repo, ask
@wkentaro, @aginika, @garaemon

--
◉ Kei Okada

**Comment by wkentaro** [comment]
this one?
http://jsk-docs.readthedocs.org/en/latest/jsk_visualization/doc/jsk_rqt_plugins/plugins/rqt_histogram_plot.html
